File: engines/python/src/numbrane_python/seed_streams.py
"""Named deterministic RNG substreams from a NAP u32 seed.

Algorithm (must match Rust ``numbrane_core::seed_streams``):

1. Hash the UTF-8 stream name with **FNV-1a 32-bit**
   (offset basis ``0x811C9DC5``, prime ``0x01000193``).
2. Mix with the recipe seed: ``state = (seed ^ h) & 0xFFFFFFFF``.
3. Advance **one** splitmix32 step (same constants as ``rng._splitmix32``)
   and use that output as the stream seed.

Derivation is order-independent: each name maps to a fixed u32 regardless of
which other streams are requested.
"""

from __future__ import annotations

from numbrane_python.rng import MASK, _splitmix32

STREAM_NAMES: tuple[str, ...] = (
    "geometry",
    "field",
    "particles",
    "fractal",
    "palette",
    "audio",
    "interaction",
)

FNV_OFFSET = 0x811C9DC5
FNV_PRIME = 0x01000193


def fnv1a32(data: bytes) -> int:
    """FNV-1a 32-bit hash over raw bytes."""
    h = FNV_OFFSET
    for b in data:
        h ^= b
        h = (h * FNV_PRIME) & MASK
    return h


def stream_seed(seed: int, name: str) -> int:
    """Derive one named substream seed from a NAP u32 recipe seed."""
    h = fnv1a32(name.encode("utf-8"))
    state = [(int(seed) ^ h) & MASK]
    return _splitmix32(state)


def derive_streams(seed: int, names: tuple[str, ...] = STREAM_NAMES) -> dict[str, int]:
    """Return ``{stream_name: u32}`` for each named substream."""
    seed &= MASK
    return {name: stream_seed(seed, name) for name in names}

File: engines/web/src/latticefall/seedStreams.ts
/**
 * Named deterministic RNG substreams — must match Python/Rust.
 *
 * Algorithm:
 * 1. FNV-1a 32 over UTF-8 stream name
 * 2. state = (seed ^ h) & 0xffffffff
 * 3. one splitmix32 step → stream seed
 */

const MASK = 0xffff_ffff;
const FNV_OFFSET = 0x811c_9dc5;
const FNV_PRIME = 0x0100_0193;

export const STREAM_NAMES = [
  "geometry",
  "field",
  "particles",
  "fractal",
  "palette",
  "audio",
  "interaction",
] as const;

export type StreamName = (typeof STREAM_NAMES)[number];

function splitmix32(state: number): number {
  let z = (state + 0x9e37_79b9) >>> 0;
  z = Math.imul(z ^ (z >>> 16), 0x85eb_ca6b) >>> 0;
  z = Math.imul(z ^ (z >>> 13), 0xc2b2_ae35) >>> 0;
  return (z ^ (z >>> 16)) >>> 0;
}

export function fnv1a32(data: Uint8Array | string): number {
  const bytes =
    typeof data === "string" ? new TextEncoder().encode(data) : data;
  let h = FNV_OFFSET;
  for (let i = 0; i < bytes.length; i++) {
    h ^= bytes[i]!;
    h = Math.imul(h, FNV_PRIME) >>> 0;
  }
  return h >>> 0;
}

export function streamSeed(seed: number, name: string): number {
  const h = fnv1a32(name);
  const state = ((seed >>> 0) ^ h) >>> 0;
  return splitmix32(state);
}

export function deriveStreams(
  seed: number,
  names: readonly string[] = STREAM_NAMES,
): Record<string, number> {
  const s = seed >>> 0;
  const out: Record<string, number> = {};
  for (const name of names) {
    out[name] = streamSeed(s, name);
  }
  return out;
}
